app/controllers/concerns/admin_only_maintenance_mode.rb

```diff
@@ -0,0 +1,21 @@
+module AdminOnlyMaintenanceMode
+  extend ActiveSupport::Concern
+
+  included do
+    before_action :redirect_non_admins_in_maintenance_mode
+  end
+
+private
+
+  def redirect_non_admins_in_maintenance_mode
+    return unless Settings.features.maintenance_mode.enabled
+
+    flash[:notice] = Settings.features.maintenance_mode.message
+
+    return if %w[sign_in sessions].include?(controller_name)
```

Comment:
This is probably fine for now but I wonder if it would be better to have something that we add to the actual pages that are allowed in maintenance mode. Make it more explicit and visible on the controller rather than buried in here. This module gets included in `ApplicationController` so we'd be able to implement it in here and have it available within all of the controllers.

Reply:
I was thinking it would be good to contain everything related to this feature inside this module. That way we don't have "if maintenance on" in multiple locations.